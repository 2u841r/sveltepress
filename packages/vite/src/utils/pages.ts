import fg from 'fast-glob'

export async function getPages() {
  return (await fg('src/routes/**/+page.(md|svelte)'))
}
