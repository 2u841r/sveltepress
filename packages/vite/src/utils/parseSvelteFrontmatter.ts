import { parse } from 'svelte/compiler'
import type { Expression, Pattern, PrivateIdentifier, SpreadElement } from 'estree'

/**
 * Parse the svelte source and get the frontmatter
 * For example
 * ```svelte
 * <script context="module">
 *   export const frontmatter = {
 *     title: 'some title',
 *     foo: 'bar'
 *   }
 * </script>
 * ```
 * Code like this would parsed this result
 * ```js
 * {
 *   title: 'some title',
 *   foo: 'bar'
 * }
 * ```
 * @param svelteCode the svelte source code
 */
export function parseSvelteFrontmatter(svelteCode: string) {
  const fm: Record<string, any> = {}
  const ast = parse(svelteCode)
  if (!ast.module)
    return fm
  ast.module.content.body.forEach((line) => {
    if (line.type !== 'ExportNamedDeclaration')
      return
    if (line.declaration?.type !== 'VariableDeclaration')
      return
    if (line.declaration.kind !== 'const')
      return
    const variable = line.declaration.declarations[0]
    if (!variable || variable.type !== 'VariableDeclarator')
      return
    if (variable.id.type !== 'Identifier')
      return
    if (variable.id.name !== 'frontmatter')
      return
    if (!variable.init)
      return
    if (variable.init.type !== 'ObjectExpression')
      return
    variable.init.properties.forEach((prop) => {
      if (prop.type !== 'Property')
        return
      recursivelySetValue(fm, prop.key, prop.value)
    })
  })

  return fm
}

function recursivelySetValue(
  fm: Record<string, any>,
  key: Expression | PrivateIdentifier,
  value: Expression | Pattern | SpreadElement | null,
) {
  if (key.type !== 'Literal' && key.type !== 'Identifier')
    return
  const keyName = String(key.type === 'Identifier' ? key.name : key.value)
  switch (value?.type) {
    case 'Literal':
      fm[keyName] = value.value
      break
    case 'ArrayExpression':
      fm[keyName] = []
      value.elements.forEach((ele, i) => {
        recursivelySetValue(fm[keyName], { type: 'Literal', value: i }, ele)
      })
      break
    case 'ObjectExpression':
      fm[keyName] = {}
      value.properties.forEach((prop) => {
        if (prop.type === 'Property')
          recursivelySetValue(fm[keyName], prop.key, prop.value)
      })
  }
}
