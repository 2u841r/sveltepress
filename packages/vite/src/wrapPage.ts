import { resolve } from 'path'
import { writeFileSync } from 'fs'
// import LRUCache from 'lru-cache'
import fsExtra from 'fs-extra'
import type { SiteConfig } from './types'
import { BASE_PATH } from './plugin.js'

// TODO: use cache to avoid frequently IO
// const cache = new LRUCache<string, any>({ max: 1024 })

export function wrapPage({ id, svelteCode, pageLayout, fm = {}, siteConfig }: {
  pageLayout?: string
  fm?: Record<string, any>
  siteConfig: SiteConfig
  svelteCode: string
  id: string
}) {
  if (!pageLayout)
    return svelteCode

  const wrap = (sveltePagePath: string) => `<script>
    import Page from '${sveltePagePath}'
    import PageLayout from \'${pageLayout}\'
  
    const fm = ${JSON.stringify(fm)}
    const siteConfig = ${JSON.stringify(siteConfig)}
  </script>
  <PageLayout {fm} {siteConfig}>
    <Page />
  </PageLayout>
    `
  // src/routes/foo/+page.(md|svelte) => // .sveltepress/pages/foo/_page.svelte
  let routeId = id.slice(id.indexOf('/src/routes/')).replace(/^\/src\/routes/, '')

  if (id.endsWith('.md'))
    routeId = id.slice(id.indexOf('/src/routes/')).replace(/^\/src\/routes/, '').replace(/\+page.md$/, '_page.svelte')

  const fullPagePath = resolve(BASE_PATH, `pages${routeId}`)
  fsExtra.ensureFileSync(fullPagePath)
  writeFileSync(fullPagePath, svelteCode)
  return wrap(fullPagePath)
  return svelteCode
}

