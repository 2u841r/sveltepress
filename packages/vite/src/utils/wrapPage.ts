import { resolve } from 'path'
import { writeFileSync } from 'fs'
import LRUCache from 'lru-cache'
import fsExtra from 'fs-extra'
import type { MdsvexOptions } from 'mdsvex'
import type { ResolvedTheme, SiteConfig } from '../types'
import { BASE_PATH } from '../plugin.js'
import mdToSvelte from '../markdown/mdToSvelte.js'
import { parseSvelteFrontmatter } from './parseSvelteFrontmatter.js'
import { info } from './log.js'

const cache = new LRUCache<string, any>({ max: 1024 })

export async function wrapPage({ id, mdOrSvelteCode, theme, siteConfig }: {
  theme?: ResolvedTheme
  siteConfig: SiteConfig
  mdOrSvelteCode: string
  id: string
}) {
  if (!theme?.pageLayout)
    return mdOrSvelteCode

  const cacheKey = JSON.stringify({ id, mdOrSvelteCode })
  let cached = cache.get(cacheKey)
  if (cached)
    return cached

  const mdsvexOptions: MdsvexOptions = {
    highlight: {
      highlighter: theme?.highlighter,
    },
    remarkPlugins: theme?.remarkPlugins,
    rehypePlugins: theme?.rehypePlugins,
  }
  let fm: Record<string, any> = {}
  let svelteCode = ''

  // /src/routes/foo/+page.(md|svelte) => /foo/+page.(md|svelte)
  const relativeRouteFilePath = id.slice(id.indexOf('/src/routes/')).replace(/^\/src\/routes/, '')

  const routeId = relativeRouteFilePath.replace(/\+page.(md|svelte)$/, '')
  info('parsing: ', routeId)

  if (id.endsWith('.md')) {
    const { code, data } = await mdToSvelte({
      mdContent: mdOrSvelteCode,
      filename: id,
      mdsvexOptions,
    }) || { code: '', data: {} }
    fm = data?.fm || {}
    svelteCode = code
  }
  if (id.endsWith('.svelte')) {
    fm = parseSvelteFrontmatter(mdOrSvelteCode)
    svelteCode = mdOrSvelteCode
  }

  const wrap = (sveltePagePath: string) => `<script>
    import Page from '${sveltePagePath}'
    import PageLayout from \'${theme.pageLayout}\'
  
    const fm = ${JSON.stringify(fm)}
    const siteConfig = ${JSON.stringify(siteConfig)}
  </script>
  <PageLayout {fm} {siteConfig}>
    <Page />
  </PageLayout>
  `

  // src/routes/foo/+page.(md|svelte) => .sveltepress/pages/foo/_page.svelte
  // NOTICE: cannot use +page as filename cause it would case circular parse
  const routePath = relativeRouteFilePath.replace(/\+page.(md|svelte)$/, '_page.svelte')

  const fullPagePath = resolve(BASE_PATH, `pages${routePath}`)
  fsExtra.ensureFileSync(fullPagePath)
  writeFileSync(fullPagePath, svelteCode)
  const wrappedCode = wrap(fullPagePath)
  cached = {
    wrappedCode,
    fm,
    routeId,
  }
  cache.set(cacheKey, cached)
  return cached
}

