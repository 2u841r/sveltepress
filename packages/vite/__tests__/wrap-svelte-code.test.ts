import { describe, expect, it } from 'vitest'
import { wrapSvelteCode } from '../src/utils/wrapPage'

const source = `<script>
  import Foo from './Foo.svelte'
</script>

<h1>Some content</h1>

<script context="module">
  const a = 'a'
</script>

<h1>Some other content</h1>

<style>
  h1 {
    color: blue;
  }
</style>
`

describe('wrap svelte code', () => {
  it('with script, style and context script', () => {
    expect(wrapSvelteCode({
      svelteCode: source,
      fm: {
        title: 'title',
      },
      siteConfig: {
        title: 'Site title',
        description: 'Site desc',
      },
      pageLayout: '/path/to/PageLayout.svelte',
    })).toMatchInlineSnapshot(`
      "
        <script>
      import PageLayout from '/path/to/PageLayout.svelte'
      const fm = {\\"title\\":\\"title\\"}
      const siteConfig = {\\"title\\":\\"Site title\\",\\"description\\":\\"Site desc\\"}
        import Foo from './Foo.svelte'
      </script>
      <script context=\\"module\\">
        const a = 'a'
      </script>
      <PageLayout {fm} {siteConfig}>
        

      <h1>Some content</h1>



      <h1>Some other content</h1>



      </PageLayout>
      <style>
        h1 {
          color: blue;
        }
      </style>
      "
    `)
  })
})
