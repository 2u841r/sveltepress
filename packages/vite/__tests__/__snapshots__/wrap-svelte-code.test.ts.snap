// Vitest Snapshot v1

exports[`wrap svelte code > with script, style and context script 1`] = `
"<script context=\\"module\\">
import PageLayout from '/path/to/PageLayout.svelte'
const fm = {\\"title\\":\\"title\\"}
const siteConfig = {\\"title\\":\\"Site title\\",\\"description\\":\\"Site desc\\"}
  export const someVar = 'some value'
</script>
<script>
import SomeComponent from '/path/to/SomeComponent.svelte'

let innerWidth
</script>
<svelte:head>
  <style>
    .foo {
      color: red;
    }
  </style>
  <script>
    var map = null
  </script>
</svelte:head>
<svelte:body on:scroll></svelte:body>
<svelte:window bind:innerWidth={innerWidth}></svelte:window>
<PageLayout {fm} {siteConfig}>







# Title Level 1

Paragraph

<SomeComponent />




</PageLayout>
<style>
  h1 {
    color: blue;
  }
</style>
"
`;
